### PR TITLE
Remove CHIRRTL

### PR DESCRIPTION
**Pull request by azidar**
Triage of problems poised by #727 

Questions to discuss:
- [ ] Removing WIR. Removing CHIRRTL requires handling memories and the WIR stuff. I think since they are separable, this PR should only handle the memory stuff.

TODO List:
- [ ] Infer memory port direction (done in ResolveFlows?)
- [ ] Check memory port direction (done in CheckFlows?)
- [ ] Add to ir: DefMemoryPort, MDir
- [ ] Update InferTypes to know "hidden" fields of memory port: en and data (? and mask)
- [ ] Handle masks (are they also a secret hidden field?)
- [ ] Update what Chisel generates (needs to assign to hidden fields)
- [ ] Add support for DefMemoryPort in all HighForm passes and any passes prior to expand whens
  - [ ] Dedup?
  - [ ] passes.CheckHighForm,
  - [ ] passes.ResolveKinds,
  - [ ] passes.InferTypes,
  - [ ] passes.CheckTypes,
  - [ ] passes.Uniquify,
  - [ ] passes.ResolveKinds,
  - [ ] passes.InferTypes,
  - [ ] passes.ResolveFlows,
  - [ ] passes.CheckFlows,
  - [ ] new passes.InferBinaryPoints(),
  - [ ] new passes.TrimIntervals(),
  - [ ] new passes.InferWidths,
  - [ ] passes.CheckWidths,
  - [ ] new firrtl.transforms.InferResets
  - [ ] passes.PullMuxes,
  - [ ] passes.ReplaceAccesses,
  - [ ] passes.ExpandConnects,
  - [ ] passes.RemoveAccesses,
  - [ ] passes.ExpandWhens,
  - [ ] passes.CheckInitialization,
  - [ ] new checks.CheckResets,

### CODE REVIEW

**Comment by albert-magyar** [comment]
With the idea being that any transform will declare its dependencies, would it be reasonable to change `RemoveCHIRRTL` to `RemoveInferredMems` or something like that, and just have every `HighForm` pass that doesn't handle those statements declare it as a dependency?

**Comment by seldridge** [comment]
@albert-magyar: yeah, you can do it that way.

More info: The dependency-api-2 branch is simpler in that almost every later transform is specified as depending on all earlier transforms (primarily to force the old order). However, it makes more sense to relax this to accurately specify the true dependencies. Divining the true dependencies may take some work, however.

**Comment by albert-magyar** [comment]
@seldridge that set of baseline dependencies does sound very useful for "blessing" `CDefMemory` and `CDefMPort`. I imagine the following would work and remove the burden of supporting inferred memories in any passes declaring the new name of `RemoveCHIRRTL` as a dependency.

1. Compiler starts with input FIRRTL file that may contain inferred memories, inferred ports, and inferred read/write enables.
2. `CheckHighForm` (already essentially identical to `CheckCHIRRTL` since #1148).
3. `InferTypes` (would need to be merged with `CInferTypes`).
4. `RemoveCHIRRTL` (bugfixed and renamed to `RemoveInferredMems` or something). It would be possible to fix all buggy cases, but the read-enable semantics would need to be set in stone in the spec first.
5. Rest of compiler proceeds as normal.

This would also avoid any changes or special-casing in `ExpandWhens`.

File: spec/spec.tex [hunk 1]
```diff
@@ -842,6 +842,85 @@ \subsubsection{Write Under Write Behaviour}
...
+write mport w = syncmem[cmd.addr], clk
+when (cmd.wen) :
+  w <= cmd.wdata ; conditional write operation
+
+; last-connect semantics apply to inferred writes
+when (cmd.wen && cmd.isClear) :
+  w <= UInt<32>(``h00'') ; different condition
+\end{lstlisting}
+
+Inferred write ports do not support masked writes; therefore, all write operations inherently have
```
Comment (azidar) [inline spec/spec.tex:918]: Aren't masked writes inferred though?

File: src/main/antlr4/FIRRTL.g4 [hunk 2]
```diff
@@ -94,8 +94,6 @@ stmt
   : 'wire' id ':' type info?
   | 'reg' id ':' type exp ('with' ':' reset_block)? info?
   | 'mem' id ':' info? INDENT memField* DEDENT
-  | 'cmem' id ':' type info?
-  | 'smem' id ':' type ruw? info?
```
Comment (azidar) [inline src/main/antlr4/FIRRTL.g4:98]: I thought we were going to keep around the CHIRRTL syntax for backwards compatibility?

File: src/main/antlr4/FIRRTL.g4 [hunk 3]
```diff
@@ -145,8 +143,7 @@ info
   ;
 
 mdir
-  : 'infer'
```
Comment (azidar) [inline src/main/antlr4/FIRRTL.g4:148]: Why are we getting rid of `infer`? Doesn't Chisel generate this?

**Pull request closed without merging**